### ISSUE TITLE
<fanartres> in advancedsettings.xml is broken

### ISSUE THREAD

**Issue by roidy** [xbmc/xbmc#17739]
### Describe the bug
Here is a clear and concise description of what the problem is:

< fanartres > in advancedsettings.xml will never to honored due to a bug in pictures/Picture.cpp line 218

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

When setting < fanartres > to 9999 fanart should not be resized as stated in the wiki.

## Actual Behavior

Fanart is resize to < imageres > not < fanartres >

## Possible Fix

https://github.com/xbmc/xbmc/blob/0fcc3121a7c9b501631c656db5143ea4a0a2f3cf/xbmc/pictures/Picture.cpp#L214-L222

At line 218 `height` will never be greater than `advancedSettings->m_fanartRes` when < fanartres > is set to 9999 therefore line 220 will never execute and the cached image will always use < imageres >

A simple fix should be:-
line 218:
`(height >= advancedSettings->m_fanartRes || advancedSettings->m_fanartRes == 9999))`

### To Reproduce

1. In advancedsettings.xml set < imageres > to 720 and < fanartres > to 9999
2. Looking in thumbnails directory all fanart is resized to 720 pixels height 

### Debuglog

The debuglog can be found here: 
https://pastebin.com/vfqFkP3p

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: CoreELEC
 - Kodi version: 18.6

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by DaVukovic** [comment]
If you have a fix at hand, could you please PR it?

**Comment by roidy** [comment]
Yep, I'll test the changes and create a PR in the next day or two.

**Comment by roidy** [comment]
Unfortunately my proposed change doesn't seem to fix this issue.

Here a detailed description of the current problem.

Take 3 movies A, B and C. 

Movie A has fanart that is 720 pixels high.
Movie B has fanart that is 1080 pixels high.
Movie C has fanart that is 2160 pixels high.

Tested using Kodi 18.6 and Kodi 19 nightly Windows versions and CoreELEC Kodi 18.6


Setting < fanartres > to 1080 in advancedsettings.xml
 |Movie|Original fanart size|Expected Cached size|Actual Cached size Windows|Actual Cached size CoreELEC|
|--------|-------|------|------|------|
|A|1280x720|1280x720|1280x720 - OK|1280x720 - OK|
|B|1920x1080|1920x1080|1920x1080 - OK|1920x1080 - OK|
|C|3840x2160|1920x1080|1920x1080 - OK|1920x1080 - OK|


Setting < fanartres > to 2160 in advancedsettings.xml
 |Movie|Original fanart size|Expected Cached size|Actual Cached size Windows|Actual Cached size CoreELEC|
|--------|-------|------|------|----|
|A|1280x720|1280x720|1280x720 - OK|1280x720 - OK|
|B|1920x1080|1920x1080|1920x1080 - OK|1280x720 - WRONG|
|C|3840x2160|3840x2160|1920x1080 - WRONG|3840x2160 - OK|

Setting < fanartres > to 9999 in advancedsettings.xml
 |Movie|Original fanart size|Expected Cached size|Actual Cached size Windows|Actual Cached size CoreELEC|
|--------|-------|------|------|-----|
|A|1280x720|1280x720|1280x720 - OK|1280x720 - OK|1280x720 - OK|
|B|1920x1080|1920x1080|1920x1080 - OK|1280x720 - WRONG|
|C|3840x2160|3840x2160|1920x1080 - WRONG|1280x720 - WRONG|

Clearly there is something wrong with the fanartres code, I just can't figure out what it is, it's needs somebody with more knowledge of the image resizing and caching code to take a look at it.

Thanks

**Comment by KarellenX** [comment]
I use `<fanartres>9999</fanartres>` and `<imageres>9999</imageres>` on Win10 Kodi v18.6

I have not done any testing, but I searched my Thumbnails cache and sorted by width.

The quantity of 1920x1080 and 3840x2160 images in my cache are in a quantity that I would expect for the size of my library including recently scraped movies. By your example, it would mean that I should have neither of those image sizes.

I also checked the thumbnail cache of my testing setup of v19 and I found both sizes of images in the cache. I even have music artist images sized 6780x5000 and 7216x5412 in the cache.

I cannot confirm this issue on Windows

**Comment by roidy** [comment]
> I use `<fanartres>9999</fanartres>` and `<imageres>9999</imageres>` on Win10 Kodi v18.6

Please re-read my initial post, especially this part:-

> At line 218 height will never be greater than advancedSettings->m_fanartRes when < fanartres > is set to 9999 therefore line 220 will never execute and the cached image will always use < imageres >

Try the following rename your Thumbnail directory and set < imageres > to 720 and then regenerate all the thumbs and compare the sizes then.

Looking at the code block I posted above you can see that the `if` statement at line  217,218
```
if (fabsf(static_cast<float>(width) / static_cast<float>(height) / (16.0f / 9.0f) - 1.0f) <= 0.01f && 
       height >= advancedSettings->m_fanartRes)
```
Taking a 4k fanart where height is 2160 this value will never be higher than `advancedSettings->m_fanartRes` 9999 in this case 

Therefore line 220
`max_height = advancedSettings->m_fanartRes`; 

max_height will never be set to < fanartres > and will continue to be < imageres > as set in line 214.

Which works perfectly for you as you've set < imageres > to 9999 anyway..... But for people who use a lower value for < imageres > or keep it at the default of 720 then fanart will be resized to this value and never < fanartres >

**Comment by howie-f** [comment]
@roidy please give this a shot:

```
  uint32_t max_height = advancedSettings->m_imageRes;
  if (advancedSettings->m_fanartRes > advancedSettings->m_imageRes)
  { // 16x9 images larger than the fanart res use that rather than the image res
    if (fabsf(static_cast<float>(width) / static_cast<float>(height) / (16.0f / 9.0f) - 1.0f) <= 0.01f &&
        (height >= advancedSettings->m_fanartRes || advancedSettings->m_fanartRes == 9999))
       if (advancedSettings->m_fanartRes == 9999)
         max_height = height // use original height...
       else
         max_height = advancedSettings->m_fanartRes; // use height defined in fanartRes
  }
```

untested but maybe we should use the original height if fanartRes = 9999
hope it helps

**Comment by roidy** [comment]
This is the fix I've come up with:-

Picture.cpp
```
  uint32_t max_height = advancedSettings->m_imageRes;
  if (advancedSettings->m_fanartRes > advancedSettings->m_imageRes)
  { // 16x9 images larger than the fanart res use that rather than the image res
    if (fabsf(static_cast<float>(width) / static_cast<float>(height) / (16.0f / 9.0f) - 1.0f) <= 0.01f)
    {
      max_height = advancedSettings->m_fanartRes; // use height defined in fanartRes
    }
  }
  
  uint32_t max_width = max_height * 16/9;

  dest_height = std::min(dest_height, max_height);
  dest_width  = std::min(dest_width, max_width);
```

No need to set to original height as the value is clamped to the minimum value in the last two lines. I've built and tested the changes and they seem to work fine.

**Comment by howie-f** [comment]
closed by #17824 - someone hit the button, please?
